### PR TITLE
Navigation: Use gap instead of margin.

### PR DESCRIPTION
**Pull request by jasmussen**
## Description

This PR refactors the navigation block to use `gap` instead of margins for spacing items out. The rules remain the same, if a navigation block has a background color, it has a smaller gap, than if it doesn't.

Submenu items should be unaffected, as those are not spaced out using margins at all, purely the padding of items themselves.

This comes with a few benefits:

- If nav items wrap, the right-most item on row 1 doesn't have margin on its right.
- The CSS is vastly simplified and reduced, making it easier for themes to override.
- Paired with #32366, this could provide a neat interface directly in the navigation block for controlling item-spacing. 

## Testing instructions

Compare this branch with trunk, and the spacing should be the same. Before:

before

After:

after

Please test with and without backgrounds, vertical and horizontal, frontend and backend. Test a few themes. Some themes like TwentyTwentyOne provide style overrides for the navigation block that are obsolete, so won't necessarily be accurate, see also https://core.trac.wordpress.org/ticket/53220.

## Screenshots

Screenshot 2021-06-01 at 12 25 32

Screenshot 2021-06-01 at 12 25 45

colors

Screenshot 2021-06-01 at 12 28 00

Screenshot 2021-06-01 at 12 28 12

Screenshot 2021-06-01 at 12 28 16

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Review by mkaz — COMMENTED** [review]
This tests well with just link menus. A minor concern would be the [Can I Use stats](https://caniuse.com/?search=gap) it is available in most desktop and current mobile, but the previous Safari still show 11% usage.

Also, the margins are off for social links, you can see in the GIF they are up against the border while the links are not. This looks the same without this change, so not a regression, so if an issue can be addressed elsewhere.
 
gap-social

My biggest question for approving the change is lack of support in the older mobile Safari browser.

**Comment by jasmussen** [comment]
Thanks for looking. Valid point about social links — and ultimately I think we should move those towards using gap as well: it's a much simpler way of ensuring harmonious spacing without an escalating set of width and nth-child rules for margins around wrapping elements.

In this particular case, the before/after should visually look identical, with and without gap for social links:

Screenshot 2021-07-01 at 10 59 34

The other part of your question is a good one: can we start to use gap more widely considering many are still using Safari 13? Utimately yes, I think so, because what happens if gap is not supported is a purely decorative issue, and not functional. So if not now, then at least soon. We landed this change for Buttons, making the same evaluation of complexity vs. legacy support. Navigation is perhaps a bit more prominent, with lack of support being potentially more noticable. 

The other half of the equation is global styles support, as tracked in https://github.com/WordPress/gutenberg/issues/32366. Right now you _can_ actually change margins using global styles, but since you can't change the gap, arguably it would remove that feature (even if currently not used at all). Perhaps I should mark this one as blocked until https://github.com/WordPress/gutenberg/issues/32366 is addressed, and then we can revaluate the browser support situation.

**Comment by jasmussen** [comment]
Since #32659 has landed, this is now unblocked and can be picked up again, so I rebased it. One thing I noticed though, is that there's a bit of a markup discrepancy when combining the nav menu with social links. Here's the editor:

Screenshot 2021-08-09 at 09 09 48

Note how there's `gap` between all nav block items, and social links is treated as an item. This is the markup:

Screenshot 2021-08-09 at 09 10 42

Note how the social links is a descendant of `wp-block-navigation__container`.

Here's the frontend:

Screenshot 2021-08-09 at 09 10 58

Here, the gap doesn't affect the Social Links block at all, and the answer lies in the markup:

Screenshot 2021-08-09 at 09 11 05

Here, Social Links is a descendant directly of `wp-block-navigation`, not the container inside.

While this is something I can work with, I'm not sure this discrepancy is intentional, and if we could fix it it would be nice. @tellthemachines @draganescu do you know if this is intentional or something I could/should ticket?

**Comment by tellthemachines** [comment]
> While this is something I can work with, I'm not sure this discrepancy is intentional, and if we could fix it it would be nice.

It is intentional, a result of the decision to decouple the front end from the editor markup in the Navigation block. The divergence is exactly because we need to wrap the Nav link items in a `ul` and we couldn't do so in the editor (or at least we can't without extensive changes to how inner blocks work, and we decided to do it this way instead because the markup differences were relatively minor).

For the styling to work across the two sets of markup, can we ignore the `ul` or `div` altogether and style the inner blocks as if they weren't wrapped? 

If it's any help, we could remove the `wp-block-navigation__container` class from the div wrapper in the editor.

**Comment by jasmussen** [comment]
> It is intentional, a result of the decision to decouple the front end from the editor markup in the Navigation block. The divergence is exactly because we need to wrap the Nav link items in a ul and we couldn't do so in the editor (or at least we can't without extensive changes to how inner blocks work, and we decided to do it this way instead because the markup differences were relatively minor).

Ah, gotcha. I just meant that social links could still have been inside the `wp-block-navigation__container`, just needing a wrapping `li` of course.

> For the styling to work across the two sets of markup, can we ignore the ul or div altogether and style the inner blocks as if they weren't wrapped?
> If it's any help, we could remove the wp-block-navigation__container class from the div wrapper in the editor.

Yes, I think that's an option. Let me dive into the markup once more with this helpful context, and see what might work best 👍 👍

**Comment by jasmussen** [comment]
Dove in for a bit, and formed two thoughts. First off, it does seem like the `wp-block-navigation__container` is not supremely useful as a wrapping div. It doesn't seem to serve an important purpose. However it also doesn't seem like it would solve the discrepancy; the idea being that the social links block is a block just like a menu item, so ideally they're both children of the same container. Can we do that? I.e. make the social links block a child of the `wp-block-navigation__container` UL, like so:

Screenshot 2021-08-11 at 10 33 46

In quick debugging it seems like it would work and be valid:

diff

It's entirely possible I'm missing a markup nuance though, and I can work with the existing markup if need be, let me know!

**Comment by tellthemachines** [comment]
> First off, it does seem like the `wp-block-navigation__container` is not supremely useful as a wrapping div. It doesn't seem to serve an important purpose.

That div is only there on the editor side because we need an element that's not the outer `nav` wrapper to attach the inner blocks logic to; it doesn't serve any purpose visually. We need a div in that place, but we don't need it to have a classname, so we can remove that at least.

> Can we do that? I.e. make the social links block a child of the wp-block-navigation__container UL

Sure, we can do that. My only reservation is that, though it's semantically correct, it might not make sense in terms of the content. 

The only other nested `ul`s we have are submenus, and those have a top level item that specifies their place as sub-lists of a main list, whereas the social icons wouldn't have that. As content, they are unlikely to be related to the other menu items, so it's hard to tell why they should be a sub-list. The fact that social icons is its own block already points to that, otherwise we might have a social icon menu item type allowing us to intersperse individual social links with other menu items. 

And due to the way we wrap menu items in `ul`s on the server side, we could easily end up with something nonsensical like this:

```
<nav class="wp-block-navigation">
  <ul class="wp-block-navigation__container">
    <li class="wp-block-navigation-link">...</li>
  </ul>
  <form role="search">...</form>
  <ul class="wp-block-navigation__container">
    <li>
      <ul class="wp-block-social-links">
        <li class="wp-block-social-link">...</li>
      </ul>
    </li>
  </ul>
</nav>
```

Having said that, if it does make it a lot easier to style the nav consistently, we should still consider doing it.

**Comment by jasmussen** [comment]
All excellent points! I think I'll pause this one until the new unified CSS classes have landed. The followup refactor will likely make it much more clear what the best path forward is. Thanks for your input 👌

**Comment by jasmussen** [comment]
Rebased this one and fixed an issue also outlined in #34351.

File: packages/block-library/src/navigation/style.scss
Comment (tellthemachines) [inline packages/block-library/src/navigation/style.scss]: On the front end, `wp-block-page-list` isn't inside `wp-block-navigation__container`, so this won't apply. I think we could go with `.wp-block-navigation .wp-block-page-list` instead.

File: packages/block-library/src/navigation/style.scss
Comment (tellthemachines) [inline packages/block-library/src/navigation/style.scss]: Same as above.

**Review by tellthemachines — COMMENTED** [review]
Rebased this PR so we can test with the recently-merged #33316. It's working well! The only part that isn't working is the Pages List styles; I left comments on that below. 

Also, I now see why you were apprehensive about the Social Icons markup:
Screen Shot 2021-09-01 at 1 58 57 pm

The only way I could find of getting all these items to line up was setting everything to `display: inline`, but that won't allow justification, so not a viable solution. 

I like the idea of allowing individual social icons in the Nav, instead of being grouped in a Social Icons block. Can you think of any reasons not to try that? If not, I'm happy to give it a go in another PR 😄

**Comment by jasmussen** [comment]
Thank you for the rebase and the review. Let me address your comments and see what I can do about the page list.

**Comment by jasmussen** [comment]
Forgot to answer this one:

> I like the idea of allowing individual social icons in the Nav, instead of being grouped in a Social Icons block. Can you think of any reasons not to try that? If not, I'm happy to give it a go in another PR 😄

The only downside is that the block inserter becomes very unwieldy having items like Page Link intermixed with every social link under the sun. #34041 should mitigate that, especially if paired with some way to curate the quick inserter. I would definitely agree that it's a better experience to not have to have an extra wrapper just for social links.

File: packages/block-library/src/navigation/style.scss
Comment (jasmussen) [reply packages/block-library/src/navigation/style.scss]: Excellent, that worked:

Screenshot 2021-09-01 at 10 20 27

Screenshot 2021-09-01 at 10 20 36

**Comment by jasmussen** [comment]
I guess this one is mostly ready! I had some comments in https://github.com/WordPress/gutenberg/pull/33812#issuecomment-910041977 on how to better integrate with global styles, and perhaps make the navigation block a flex container as defined in block.json rather than in CSS as is done here. But that's probably best left as a followup.

**Review by tellthemachines — APPROVED** [review]
Page list is looking much better now 🎉 

There's a slight discrepancy in spacing between regular items and page list items due to the padding applied to `wp-block-navigation-link__content`, especially noticeable in the vertical nav:
Screen Shot 2021-09-02 at 10 21 52 am
But that's unrelated to these changes so let's fix it separately.

> The only downside is that the block inserter becomes very unwieldy having items like Page Link intermixed with every social link under the sun.

Good point 😄  I guess it comes down to whether it's a better experience to have all those links directly available, or to have to add an intermediate block in order to access them.

If we do want to keep using the Social Links block, another thing we could try is nesting the whole block inside the `wp-block-navigation__container` and then adding `role="presentation"` to the extra wrapping `li` and `ul` so that semantically the social links are siblings of the main list items. I haven't actually tried this yet, so we'd need to test it well to make sure it works. In terms of the layout though, having that extra wrapper in place means that individual social links won't be able to wrap like the other nav items - much like what currently happens with Page List.

**Comment by tellthemachines** [comment]
> There's a slight discrepancy in spacing between regular items and page list items due to the padding applied to wp-block-navigation-link__content, especially noticeable in the vertical nav:

Actually forget it, those are theme styles, aren't they 🤦

**Comment by jasmussen** [comment]
> Actually forget it, those are theme styles, aren't they 🤦

If you were using TT1 to test, there are some [known issues with menu item paddings and margins there](https://core.trac.wordpress.org/ticket/53220). I did some testing but couldn't reproduce, but if we find a way to reproduce I'll be happy to follow up and fix!

> Good point 😄 I guess it comes down to whether it's a better experience to have all those links directly available, or to have to add an intermediate block in order to access them.

Honestly I think your point is well made, and that the issue is worth solving. Some of the efforts outlined in #34041 will need to handle these cases anyway, and a curated quick inserter I honestly believe would benefit all blocks with nesting.

Thank you for the reviews. I think I'll land this one and follow up if anything comes up.

**Pull request merged**
